drake/core/Gradient.h

```diff
@@ -18,55 +18,158 @@ namespace Drake {
-    initTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  /** \brief initialize a single autodiff matrix given the corresponding value matrix.
+   * @param[in] mat 'regular' matrix of values
+   * @param[out] ret AutoDiff matrix
+   * @param[in] num_derivatives the size of the derivatives vector @default the size of mat
+   * @param[in] deriv_num_start starting index into derivative vector (i.e. element deriv_num_start in derivative vector corresponds to mat(0, 0)). @default 0
+   */
+  template<typename Derived, typename DerivedAutoDiff>
+  void initializeAutoDiff(const Eigen::MatrixBase<Derived> &val, Eigen::MatrixBase<DerivedAutoDiff> &auto_diff_matrix, Eigen::DenseIndex num_derivatives = -1, Eigen::DenseIndex deriv_num_start = 0) {
+    using ADScalar = typename DerivedAutoDiff::Scalar;
+    static_assert(Derived::RowsAtCompileTime == DerivedAutoDiff::RowsAtCompileTime, "auto diff matrix has wrong number of rows at compile time");
+    static_assert(Derived::ColsAtCompileTime == DerivedAutoDiff::ColsAtCompileTime, "auto diff matrix has wrong number of columns at compile time");
+
+    if (num_derivatives == -1)
+      num_derivatives = val.size();
+
+    auto_diff_matrix.resize(val.rows(), val.cols());
+    Eigen::DenseIndex deriv_num = deriv_num_start;
+    for (int i = 0; i < val.size(); i++) {
+      auto_diff_matrix(i) = ADScalar(val(i), num_derivatives, deriv_num++);
+    }
   }
 
-  /** constantTaylorVecXd
-   * @brief initializes the vector with x=val and dx=zeros(numel(val))
+  /** \brief initialize a single autodiff matrix given the corresponding value matrix.
+   * @param[in] mat 'regular' matrix of values
+   * @param[in] num_derivatives the size of the derivatives vector @default the size of mat
+   * @param[in] deriv_num_start starting index into derivative vector (i.e. element deriv_num_start in derivative vector corresponds to mat(0, 0)). @default 0
+   * @return AutoDiff matrix
    */
-  template <typename DerivedA, typename DerivedB>
-  void constantTaylorVecXd(const Eigen::MatrixBase<DerivedA>& val, Eigen::MatrixBase<DerivedB>& auto_diff_vec) {
-    for (int i=0; i<val.size(); i++) {
-      auto_diff_vec(i) = Eigen::AutoDiffScalar<typename DerivedB::Scalar::DerType>(val(i));
-    }
+  template<int Nq = Eigen::Dynamic, typename Derived>
+  AutoDiffMatrixType<Derived, Nq> initializeAutoDiff(const Eigen::MatrixBase<Derived> &mat, Eigen::DenseIndex num_derivatives = -1, Eigen::DenseIndex deriv_num_start = 0) {
+    if (num_derivatives == -1)
+      num_derivatives = mat.size();
+
+    AutoDiffMatrixType<Derived, Nq> ret(mat.rows(), mat.cols());
+    initializeAutoDiff(mat, ret, num_derivatives, deriv_num_start);
+    return ret;
   }
 
-  template <typename Derived>
-  TaylorVecXd constantTaylorVecXd(const Eigen::MatrixBase<Derived>& val) {
-    TaylorVecXd auto_diff_vec(val.rows());
-    constantTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  // TODO begin: move to different location?
```

Comment:
Yeah, so is there a better place for this. At first I thought Vector.h, but I need the SizeAtCompileTime, not the RowsAtCompileTime. We could switch the Vector stuff over to using SizeAtCompileTime.